docs/proposals/pepxxx_wheel_variant_support.md

```diff
@@ -62,88 +62,650 @@ something they [indirectly do today](https://github.com/wheelnext/wheelnext/pull
+
+```
+{import path}(:{object path})?
 ```
 
-The plugin itself follows a minimal interface:
+where import path specifies the module to import, as for the Python `import` statement, and object path specifies
+the class to use. If object path is omitted, the whole module is used as the endpoint.
+
+If the API endpoint specifies a callable, it is called to instantiate the provider object. Otherwise, it is used as-is.
+
+An API endpoint specification is equivalent to the following Python pseudocode:
+
+```python
+import {import path}
+
+if {object path}:
+    obj = {import path}.{object path}
+else:
+    obj = {import path}
+
+if callable(obj):
+    obj = obj()
+```
+
+Additionally, a plugin provider can install an entry point in the `variant_plugins` group that can be used
+by development tools to discover available providers. However, wheels must be installable without the presence of entry
+points.
+
+#### Plugin API
+
+The plugin API needs to conform to the `PluginType` protocol as exemplified in the following snippet:
 
 ```python
-from variantlib.config import ProviderConfig
-from my_plugin import __version__
+class VariantFeatureConfigType:
+    name: str
+    values: list[str]
+
+
+cladd VariantPropertyType:
+    namespace: str
+    feature: str
+    value: str
+
 
-class MyVariantPlugin:
-    __provider_name__ = "my_plugin"
-    __version__ = __version__
+class PluginType:
+    namespace: str
+    dynamic: bool
 
-    def run(self) -> ProviderConfig | None:
-        """
-        Detects platform attributes and returns a ProviderConfig if applicable.
-        """
+    def get_supported_configs(
+        self, known_properties: frozenset[VariantPropertyType] | None
+    ) -> list[VariantFeatureConfigType]:
+       ...
+
+    def validate_property(self, variant_property: VariantPropertyType) -> bool:
+       ...
 ```
 
-### Integration with `installers`
+The plugin must implement the following attributes or properties:
+
+- `namespace` stating the namespace used by the provider
+
+- `dynamic` indicating whether the plugin is dynamic
+
+It must also implement two methods or functions:
+
+- `get_supported_configs()` that returns a list of feature names and values that are compatible with the current
+  environment, in their order of preference (i.e. a wheel with such a property can be installed)
+
+- `validate_property()` that checks whether the specified property name and value is valid (i.e. a wheel can be built
+  with a such a property)
+
+#### `get_supported_configs()`
+
+The `get_supported_configs()` method is used to obtain the list of configurations that are supported by the current
+environment. Its exact semantics depends on whether the provider plugin is declared as dynamic or not.
+
+If the provider is static (`dynamic = False`), the list of supported configurations is expected to be fixed.
+The `known_properties` parameter is always `None`, and the method must return all configurations supported
```

Comment:
Can we give parameter a default value of `None`?

Reply:
We could, but to be honest I don't see what that gives us. The frontend will always have to provide the correct value based on `dynamic`, so it will either not achieve anything, or be actively confusing.